cartofacade/cgo_api.go

```diff
@@ -0,0 +1,456 @@
+	}
+	sensorSlice := unsafe.Slice(pSensor, sz)
+	for i, sensor := range cfg.Sensors {
+		sensorSlice[i] = goStringToBstring(sensor)
+	}
+	lidarCfg, err := toLidarConfig(cfg.LidarConfig)
+	if err != nil {
+		return C.viam_carto_config{}, err
+	}
+
+	vcc.sensors = pSensor
+	vcc.sensors_len = C.int(sz)
+	vcc.map_rate_sec = C.int(cfg.MapRateSecond)
+	vcc.data_dir = goStringToBstring(cfg.DataDir)
+	vcc.component_reference = goStringToBstring(cfg.ComponentReference)
+	vcc.lidar_config = lidarCfg
+
+	return vcc, nil
+}
+
+func toAlgoConfig(acfg CartoAlgoConfig) C.viam_carto_algo_config {
+	vcac := C.viam_carto_algo_config{}
+	vcac.optimize_on_start = C.bool(acfg.optimizeOnStart)
+	vcac.optimize_every_n_nodes = C.int(acfg.optimizeEveryNNodes)
+	vcac.num_range_data = C.int(acfg.numRangeData)
+	vcac.missing_data_ray_length = C.float(acfg.missingDataRayLength)
+	vcac.max_range = C.float(acfg.maxRange)
+	vcac.min_range = C.float(acfg.minRange)
+	vcac.max_submaps_to_keep = C.int(acfg.maxSubmapsToKeep)
+	vcac.fresh_submaps_count = C.int(acfg.freshSubmapsCount)
+	vcac.min_covered_area = C.double(acfg.minCoveredArea)
+	vcac.min_added_submaps_count = C.int(acfg.minAddedSubmapsCount)
+	vcac.occupied_space_weight = C.double(acfg.occupiedSpaceWeight)
+	vcac.translation_weight = C.double(acfg.translationWeight)
+	vcac.rotation_weight = C.double(acfg.rotationWeight)
+	return vcac
+}
+
+func toGetPositionResponse(value C.viam_carto_get_position_response) GetPosition {
+	return GetPosition{
+		X: float64(value.x),
+		Y: float64(value.y),
+		Z: float64(value.z),
+
+		Ox:    float64(value.o_x),
+		Oy:    float64(value.o_y),
+		Oz:    float64(value.o_z),
+		Theta: float64(value.theta),
+
+		Real: float64(value.real),
+		Imag: float64(value.imag),
+		Jmag: float64(value.jmag),
+		Kmag: float64(value.kmag),
+
+		ComponentReference: bstringToGoString(value.component_reference),
+	}
+}
+
+// freeBstringArray used to cleanup a bstring array (needs to be a go func so it can be used in tests)
+func freeBstringArray(arr *C.bstring, length C.int) {
+	C.free_bstring_array(arr, length)
+}
+
+func toError(status C.int) error {
+	switch int(status) {
+	case C.VIAM_CARTO_SUCCESS:
+		return nil
+	case C.VIAM_CARTO_UNABLE_TO_ACQUIRE_LOCK:
+		return errors.New("VIAM_CARTO_UNABLE_TO_ACQUIRE_LOCK")
+	case C.VIAM_CARTO_VC_INVALID:
+		return errors.New("VIAM_CARTO_VC_INVALID")
+	case C.VIAM_CARTO_OUT_OF_MEMORY:
+		return errors.New("VIAM_CARTO_OUT_OF_MEMORY")
+	case C.VIAM_CARTO_DESTRUCTOR_ERROR:
+		return errors.New("VIAM_CARTO_DESTRUCTOR_ERROR")
+	case C.VIAM_CARTO_LIB_PLATFORM_INVALID:
+		return errors.New("VIAM_CARTO_LIB_PLATFORM_INVALID")
+	case C.VIAM_CARTO_LIB_INVALID:
+		return errors.New("VIAM_CARTO_LIB_INVALID")
+	case C.VIAM_CARTO_LIB_NOT_INITIALIZED:
+		return errors.New("VIAM_CARTO_LIB_NOT_INITIALIZED")
+	case C.VIAM_CARTO_SENSORS_LIST_EMPTY:
+		return errors.New("VIAM_CARTO_SENSORS_LIST_EMPTY")
+	case C.VIAM_CARTO_UNKNOWN_ERROR:
+		return errors.New("status code unclassified")
```

Comment:
```suggestion
		return errors.New("VIAM_CARTO_UNKNOWN_ERROR")
```

The nuance between this case & the default case, is that this case means that the C++ code threw an exception that didn't have a corresponding status code. The default case means that the C code returned a status code that the go code couldn't classify. We need to be able to distinguish between these 2 cases.